### DISCUSSION TITLE
Frequent socket closes (with code 1001)

### DISCUSSION THREAD

**Discussion by morajabi** [enisdenjo/graphql-ws#116] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/12202757/108033084-b87ba800-7048-11eb-8c75-72d312e7a87e.png)

Here's my config:

```typescript
// https://github.com/enisdenjo/graphql-ws
let restartRequestedBeforeConnected = false
let gracefullyRestart = () => {
  restartRequestedBeforeConnected = true
}

wsClient.current = createWSClient({
  url: `${baseUrl}/graphql`,
  retryAttempts: 50,
  keepAlive: ms('1 hour'),
  lazy: typeof window === 'undefined',
  webSocketImpl: typeof window === 'undefined' ? require('ws') : WebSocket,
  connectionParams: () => {
    return {
      token: getToken(),
      activeSpaceId: getActiveSpaceId(),
      appVersion: process.env.appVersion,
    }
  },

  on: {
    closed: console.error,

    connected: (socket: unknown) => {
      console.info('client connected.', socket)
      gracefullyRestart = () => {
        if (socket instanceof WebSocket) {
          if (socket.readyState === WebSocket.OPEN) {
            console.info('emitted client restart')
            socket.close(4205, 'Client Restart')
          }
        } else {
          console.error('socket not instance of websocket')
        }
      }

      // just in case you were eager to restart
      if (restartRequestedBeforeConnected) {
        restartRequestedBeforeConnected = false
        gracefullyRestart()
      }
    },
  },
})
```

I use it with URQL:
```typescript

export const client = createClient({
  url: '/graphql',
  exchanges: [
    devtoolsExchange,
    dedupExchange,
//    ...,
    subscriptionExchange({
      enableAllOperations: true,
      forwardSubscription(operation) {
        return {
          subscribe: sink => {
            const dispose = wsClient.current?.subscribe(operation, sink)

            return {
              unsubscribe: () => dispose?.(),
            }
          },
        }
      },
    }),
  ],
})
```

**Comment by enisdenjo** [comment, marked as answer]
Hey hey, the close event is triggered by the server. Client is just communicating it. Our server implementation closes with 1001 when the server is going away (or shutting down as the close reason suggests).
https://github.com/enisdenjo/graphql-ws/blob/88a12ef3d9261e787a150d226210856688bf97da/src/use/ws.ts#L126-L128

Looking at the close reason from the provided dev-tools screenshot, the server is closing the client with a `"shutting down"` reason. This tells me that you're not using the `graphql-ws` server implementation, but rather a different one. So, sadly, I cannot be of much help - you have to check your server implementation out.

**Reply by morajabi** [reply]
Turns out it was not having keep alive on the server! Thanks!

**Reply by morajabi** [reply]
Having a note in readme would be helpful on what things default server takes care of whereas we don’t have on our own?

**Reply by enisdenjo** [reply]
As mentioned in the original answer: I doubt its the built-in `graphql-ws/lib/use/ws` server implementation we are dealing here with.

Which server do you use? Or, are you implementing a server yourself?

**Reply by morajabi** [reply]
Yes your severer has keepalive but I didn't know we needed one using ws module’s server.

**Reply by enisdenjo** [reply]
I am having a hard time understanding you.

Are you implementing your own server using `import { makeServer } from 'graphql-ws'`, or are you using the built in implementation from `import { createServer } from 'graphql-ws/lib/use/ws'`, or are you using a different server altogether?

**Reply by morajabi** [reply]
I’m using makeServer and connecting it to ws package for the server.
<img width="820" alt="Show there-sothere repo" src="https://user-images.githubusercontent.com/12202757/108094985-9f004d80-7094-11eb-95ff-5fe52b4ede07.png">

**Reply by enisdenjo** [reply]
So, no, you're not using the built in WS implementation. You're implementing your own server. In this case, no need to change the readme because its up to the user how he wants to implement the server.
